src/_pytest/fixtures.py

```diff
@@ -729,6 +724,98 @@ def _schedule_finalizers(
         finalizer = functools.partial(fixturedef.finish, request=subrequest)
         subrequest.node.addfinalizer(finalizer)
 
+
+@final
+class TopRequest(FixtureRequest):
+    """The type of the ``request`` fixture in a test function."""
+
+    def __init__(self, pyfuncitem: "Function", *, _ispytest: bool = False) -> None:
+        super().__init__(
+            fixturename=None,
+            pyfuncitem=pyfuncitem,
+            arg2fixturedefs=pyfuncitem._fixtureinfo.name2fixturedefs.copy(),
+            arg2index={},
+            fixture_defs={},
+            _ispytest=_ispytest,
+        )
+
+    @property
+    def _scope(self) -> Scope:
+        return Scope.Function
+
+    @property
+    def node(self):
+        return self._pyfuncitem
+
+    def __repr__(self) -> str:
+        return "<FixtureRequest for %r>" % (self.node)
+
+    def _fillfixtures(self) -> None:
+        item = self._pyfuncitem
+        fixturenames = getattr(item, "fixturenames", self.fixturenames)
+        for argname in fixturenames:
+            if argname not in item.funcargs:
+                item.funcargs[argname] = self.getfixturevalue(argname)
+
+    def addfinalizer(self, finalizer: Callable[[], object]) -> None:
+        self.node.addfinalizer(finalizer)
+
+
+@final
+class SubRequest(FixtureRequest):
```

Comment:
we might want to split SubRequest into plain and parametrized while at it

Reply:
This would be nice indeed.

Reply:
I'll look into it and report back.

Reply:
Looked into it:

In terms of the internal implementation it doesn't add much, mostly allows us to replace a couple of `hasattr(request, "param")` with `isinstance(request, ParametrizedSubRequest)`.

In terms on the user-facing API/typing, since we're not exposing SubRequest yet it seems premature to consider (see https://github.com/pytest-dev/pytest/pull/11219#discussion_r1265486421). It needs more work I think, assuming we want to expose only a single type to the user.

Reply:
> mostly allows us to replace a couple of hasattr(request, "param") with isinstance(request, ParametrizedSubRequest).

FTR I find the latter much clearer, but I agree we can wait to decide on this.